packages/flutter_tools/lib/src/version.dart

```diff
@@ -240,11 +240,34 @@ class FlutterVersion {
   ///
   /// This function must run while [Cache.lock] is acquired because it reads and
   /// writes shared cache files.
-  Future<void> checkFlutterVersionFreshness() async {
+  ///
+  /// Based on the passed arguments, the method decides whether to perform
+  /// the version check or not.
+  /// 
+  /// - [isUpgradeCommand]: whether the called command is the upgrade command
+  /// - [versionCheckEnvironment]: the environment variable about the version
+  ///                              check
+  /// - [versionCheckFlag]: whether the version check flag is explicitly passed
+  /// - [isMachine]: whether the tool runs in a machine-only environment like CI
+  Future<void> checkFlutterVersionFreshness({
+    bool isUpgradeCommand = false,
+    bool versionCheckEnvironment = true,
+    bool versionCheckFlag = false,
+    bool isMachine = false,
+  }) async {
+    // Only perform check if sub-command is 'upgrade'
+    if (!isUpgradeCommand) {
+      return;
+    }
     // Don't perform update checks if we're not on an official channel.
     if (!kOfficialChannels.contains(channel)) {
       return;
     }
+    // Don't perform update check in case it is disabled by environment
+    // unless it is explicitly overwritten
+    if(!versionCheckEnvironment && versionCheckFlag != true){
```

Comment:
I think this can be two separate checks, each with its own comment.

Also, I think it would make sense to check `versionCheckEnvironment` first, since it is an override that will bypass all other checks (of course logically today it doesn't matter the order you check these since they're all quick, but if we change this logic in the future it might matter what order we check).

Reply:
But as far as I see, those cases can *not* be separated. It's an `&&`, not a `||`. This check only returns in case the environment variable disables version check *unless* the user explicitly overrides this behavior by passing the `--version-check` flag - for whatever reason.

Reply:
I anyway moved it to the top.